### ISSUE TITLE
Cannot add extra files to watch

### ISSUE THREAD

**Issue by patarapolw** [rollup/rollup#3414]
- Rollup Version: ^1.32.0
- Operating System (or Browser): MacOS
- Node Version (if applicable): 12.16.1
- Link to reproduction (IMPORTANT, read below): https://github.com/patarapolw/minimal-rollup-ts-pug-sass-template

### Expected Behavior

I can add extra files, not directly being part of build bundle to `watch.include`. Namely, `index.pug` and `favicon.ico`.

### Actual Behavior

I cannot make change to the watcher at all. Not even crashing it. Deleting / creating / editing files should trigger change as describe in `watch.include` array.

**Comment by whaaaley** [comment]
I was trying to switch over to "rollup only" watching tonight and discovered the same issue. Is watching non-js files possible?

**Comment by danielvy** [comment]
Yeah, that is because rollup only watches files passed through it and whose IDs has been resolved.

I add this code to every rollup script where I need to watch arbitrary files.

```
import glob from "glob";
import path from "path";

const watcher = {
  buildStart(options) {
    if (options.watch) {
      let include = options.watch && options.watch.include;
      if (!include) return;
      include = Array.isArray(include) ? include : [include];
      for (const item of include) {
        glob.sync(path.resolve(item)).forEach(filename => {
          this.addWatchFile(filename);
        });
      }
    }
  }
}
```
And then just add `watch` to my plugins.

This has been reported and requested here several times. I think this is a trivial assumption to make that often, your build depends on external files that need to trigger the build when changed. The documentation doesn't provide any hint that this isn't the case and it's up to you to figure it out by trial and error.

I don't know why this still needs to be an issue people new to rollup waste time on.

**Comment by halostatue** [comment]
@danielvy I’ve added this to my `rollup.config.js`, and it is working for the new files I want to add, but it seems to have stopped watching the _rest_ of my files.

I’m doing this with something like:

```javascript
export default {
    input: 'src/app.js',
    output: {
      dir: destination,
      entryFileNames: 'js/[name].js',
      format: 'iife',
      sourcemap: !isProduction ? 'inline' : false,
    },
    plugins: [watcher, ...plugins],
    watch: { include: ['src/**/*.graphql'] },
  }
}
```

Any clue why this might be the case? I am trying to get rollup watch to watch some GraphQL queries that get compiled into my app bundle.

**Comment by halostatue** [comment]
Nevermind. I think I see. I’m overwriting the configuration.

**Comment by halostatue** [comment]
@danielvy 

```javascript
const watcher = {
  buildStart(options) {
    if (options.watch) {
      let include = options.watch.extra
      if (!include) {
        return
      }
      include = Array.isArray(include) ? include : [include]
      for (const item of include) {
        glob.sync(path.resolve(item)).forEach((filename) => {
          this.addWatchFile(filename)
        })
      }
    }
  },
}
```

Changing the key from `include` to `extra` prevents the overwrite of the default values.

**Comment by thekip** [comment]
Shorter/cleaner version in Typescript 

```typescript
const watcher = (globs: string[]): Partial<PluginHooks> => ({
  buildStart() {
    for (const item of globs) {
      glob.sync(nodePath.resolve(item)).forEach((filename) => {
        this.addWatchFile(filename);
      });
    }
  },
});

config: RollupOptions = {
  return {
    plugins: [
      watcher(['src/**/*.graphql']),
      nodeResolve(),
      typescript({
        tsconfig: path.root('tsconfig.sw.json'),
      }),
    ],
  };
}

```

**Comment by Utkarsh-Kumar384** [comment]
> ```ts
> Partial<PluginHooks>
> ```

@thekip I tried using your plugin with rollup@2.56.2 and I can confirm that it doesn't seem to work as expected. Rebuilds don't occur even though `this.addWatchFile(filename)` receives the right path to the file. Guessing it's an issue with the rollup API itself. Not sure though just my wild guess.

**Comment by trev-dev** [comment]
> @thekip I tried using your plugin with rollup@2.56.2 and I can confirm that it doesn't seem to work as expected

I'm using `rollup@2.56.3` and things are going swimmingly. I did catch that `nodePath.resolve(item)` doesn't seem to make much sense in my context, however, as I've not renamed the `path` module to `nodePath`. Adjusting it to `path.resolve(item)` got it working for me.

**Comment by LFFATE** [comment]
> Shorter/cleaner version in Typescript
> 
> ```ts
> const watcher = (globs: string[]): Partial<PluginHooks> => ({
>   buildStart() {
>     for (const item of globs) {
>       glob.sync(nodePath.resolve(item)).forEach((filename) => {
>         this.addWatchFile(filename);
>       });
>     }
>   },
> });
> 
> const config: RollupOptions = {
>   return {
>     plugins: [
>       watcher(['src/**/*.graphql']),
>       nodeResolve(),
>       typescript({
>         tsconfig: path.root('tsconfig.sw.json'),
>       }),
>     ],
>   };
> }
> ```

`nodePath.resolve(item)` better to replace with `nodePath.resolve(__dirname, item)`

It will work better with rollup.config.js in subdirectories

**Comment by caleb531** [comment]
For everyone watching this thread: I have run into the same issue, and I plan to release a Rollup plugin soon that allows you to watch arbitrary files outside of the module graph. The implementation will be largely based off @thekip's TypeScript code (with proper attribution, of course).

Stay tuned!

**Comment by caleb531** [comment]
Okay, I have just released v1.0.0 of [rollup-plugin-watch-globs](https://www.npmjs.com/package/rollup-plugin-watch-globs). The API usage is effectively identical to how @thekip described it in [his earlier comment](https://github.com/rollup/rollup/issues/3414#issuecomment-751699335):

```js
// rollup.config.js
import watchGlobs from 'rollup-plugin-watch-globs';

export default {
  ...
  plugins: [
    watchGlobs(['src/**/*.graphql']),
    nodeResolve(),
    typescript({
      tsconfig: path.root('tsconfig.sw.json'),
    })
  ]
}
```

I hope that streamlines your Rollup config a bit more. Enjoy! 😄

**Comment by jcoffin** [comment]
Another option is to use an external tool like [Watchman](https://facebook.github.io/watchman/). That is what I do and it works well.
